Pico 3.0

PhrozenByte opened the pull request:
Since the issues with our old dependencies kinda got out of hand lately (see #528, #534 and 8a44584291685e5e163765ed6578c184c991794e), it's time to kick-start development of **Pico 3.0** :tada:

Major changes are going to be updated dependencies (Twig 3.0+ and Symfony YAML 5.0+) and lifting the requirements to PHP 7.2.5+. We definitely won't implement all of #317, but rather focus on lazy page discovery and loading. We definitely won't overhaul Pico's event system now, neither will we implement caching or i18n support with Pico 3.0 (maybe with Pico 3.1? :thinking:).

We'll also think about changing Pico's Markdown parser. Unfortunately Parsedown's development process got very unreliable lately. However, there's a lot to consider - not just BC, but also performance-wise (Parsedown is by far the most performant Markdown parser I know). Thus there's no decision made yet. Other opinions are very welcome on this topic! Also see e.g. #576

Pico 3.0 is a major release, thus it will break BC. As always, Pico's official [`PicoDeprecated` plugin](https://github.com/picocms/pico-deprecated) ([Pull Request](https://github.com/picocms/pico-deprecated/pull/2)) will do its best to maintain BC. However, be warned, we'll explicitly drop all pre-v1.0 and v1.0 behaviour that affects performance.

As always, **feedback is highly appreciated!** :heart: 

---

**Ideas**

- #581: Pico bundle package with various official plugins

---

**Release of Pico `v3.0.0-alpha.1`**

Since the mentioned issues are affecting live installations already, I decided to release a intermediate [Pico `v3.0.0-alpha.1`](https://github.com/picocms/Pico/releases/tag/v3.0.0-alpha.1). This release upgrades Pico's dependencies to Twig 2.12, Symfony YAML 3.4, Parsedown 1.7.4 and Parsedown Extra 0.8.1 for now (requiring PHP 7.0.8+). Both are still officially supported and are basically drop-in replacements, since they maintain BC on a best-effort basis ("best-effort" means "should work for most users", there can still be issues for some users). Pico `v3.0.0-alpha.1` is meant as a fast solution for all people experiencing the issues mentioned above.

**Feedback about Pico `v3.0.0-alpha.1` is highly appreciated**, especially regarding issues! :+1: 

---

**Related links**

- [List of ideas for Pico 3.0 and beyond](https://github.com/picocms/Pico/issues/317)
- [`PicoDeprecated` pull request](https://github.com/picocms/pico-deprecated/pull/2)
- [`pico-theme` pull request](https://github.com/picocms/pico-theme/pull/3)
- [`pico-composer` pull request](https://github.com/picocms/pico-composer/pull/6)
- [`composer-installer` pull request](https://github.com/picocms/composer-installer/pull/2)
- [Website pull request](https://github.com/picocms/picocms.github.io/pull/61)
- [Updated phpDoc class docs](http://picocms.org/phpDoc/pico-3.0/)

---

**Major changes**

- Require PHP 7.0.8+
- Update dependencies: Twig 2.12, Symfony YAML 3.4, Parsedown 1.7.4 and Parsedown Extra 0.8.1; this is just an interim step, we'll update to Twig 3.0+ and Symfony YAML 5.0+ later
- …

---

Resolves #528, Resolves #534 
Fixes #602, Fixes #603, Fixes #608, Fixes #623

PhrozenByte committed:
Kick-start development of Pico 3.0

PhrozenByte committed:
Update dependencies, Pico now requires PHP 7.0.8+

Pico now depends on Twig 2.12, Symfony YAML 3.4, Parsedown 1.7.4 and Parsedown Extra 0.8.1.

PhrozenByte committed:
Update Pico::VERSION and Pico::VERSION_ID

PhrozenByte committed:
Fix Travis CI trying to build Pico 3.0 on PHP 5.x

PhrozenByte committed:
Update CHANGELOG.md

PhrozenByte committed:
Version 3.0.0-alpha.1

```
* [New] Kick-start development of Pico 3.0
* [Changed] Require PHP 7.0.8+
* [Changed] Update dependencies: Twig 2.12, Symfony YAML 3.4, Parsedown 1.7.4
            and Parsedown Extra 0.8.1; this is just an interim step, we'll
            update to Twig 3.0+ and Symfony YAML 5.0+ later
```

PhrozenByte committed:
Move Pico to picocms\Pico\ namespace

PhrozenByte committed:
Update code styling to PHP 7.0+

This commit includes no functional changes.

PhrozenByte committed:
Update to Symfony YAML 5.0 and Twig 3.0

Pico now requires PHP 7.2.5+

PhrozenByte committed:
Merge branch 'master' into pico-3.0

PhrozenByte committed:
Check theme dir before trying to load theme

PhrozenByte committed:
Fix Travis CI

PhrozenByte committed:
Various small improvements

PhrozenByte committed:
Fix Travis CI

PhrozenByte committed:
Update to phpDocumentor 3 config

PhrozenByte committed:
Improve phpDocumentor 3 config

PhrozenByte committed:
Remove deprecated AbstractPlugin::__call()

Use AbstractPlugin::getPico() instead (e.g. `->getPico()->getConfig('my_config')` instead of `->getConfig('my_config')`).

PhrozenByte committed:
Remove deprecated Pico::getBaseThemeUrl()

Use Pico::getConfig('themes_url') instead.

PhrozenByte committed:
Replace substr()-based comparisons by substr_compare()

dajare commented:
If you are considering alternatives to Parsedown (and its Extra), you could have a look at [Markdown-it][m], although I don't know how well its technologies integrate with Pico's environment. I've become familiar with it on the [TopAnswers.xyz][t] platform, where it seems to be performing well.

[m]: https://github.com/markdown-it/markdown-it
[t]: https://topanswers.xyz/meta?q=72

PhrozenByte committed:
Travis CI: Use phpDocumentor 3

PhrozenByte committed:
Merge branch 'master' into pico-3.0-alpha

PhrozenByte committed:
Update CHANGELOG.md

PhrozenByte committed:
Update Pico::VERSION

Pico::VERSION_ID isn't updated because this is a alpha release.

PhrozenByte committed:
Version 3.0.0-alpha.2

```
* [Changed] Merge all fixes of Pico v2.1.2 to v2.1.4
* [Fixed] #575: Update Symfony YAML v3.4 to support PHP 8.0+
```

PhrozenByte committed:
Travis CI: Add PHP 8.0 target

PhrozenByte committed:
Add %page_*% replacements for Markdown files

`%page_id%` is replaced by the page's ID, `%page_url%` by the page's relative URL, and `%page_path%` by the dirname of page file.

PhrozenByte committed:
Remove picocms\Pico\ namespace

This reverts commit 812ae5c21548905e308f002bc6d97a042ea09aad

PhrozenByte committed:
Re-add deprecated Pico::getBaseThemeUrl() and AbstractPicoPlugin::__call()

This reverts commits efc4fb5288d39e66a81df3707255278b33506232 and bc816febfca84fed80ca96d499bb5e034b0f8e8f

PhrozenByte committed:
Code cleanup

PhrozenByte committed:
Merge branch 'master' into pico-3.0

PhrozenByte committed:
CONTRIBUTING.md: Probot Stale was replaced by GitHub Actions

PhrozenByte committed:
Revert "Added Temporary PHP 8.0 Notice"

This reverts commit ff5e345e58f6ccdb2683ea33a2fde3337853ca0c.

PhrozenByte committed:
Travis CI: Test on PHP 8.0 and 8.1

mayamcdougall commented:
@PhrozenByte, I wasn't trying to "light a fire under you" on getting this done. 🔥😓

I just wanted to offer to take some of the burden off you if you needed me to... 

(And also [pacify some of the ongoing questions](https://github.com/picocms/Pico/commit/ff5e345e58f6ccdb2683ea33a2fde3337853ca0c) people keep been asking.)

Just let me know if I've overstepped. 😔

PhrozenByte committed:
Bump to API version 4

PhrozenByte commented:
Nah, no worries, it's fine. This needs to be done and a little kick in the ass helps :sweat_smile: :see_no_evil: :wink:

As far as I remember anything should be done now code-wise. Pico is working locally again. This will be Pico 3.0 even though the only "real" change are the updated dependencies (most other changes are code cleanups and syntax updates), because it still breaks backwards compatibility (and in a "hard" way, i.e. `PicoDeprecated` can't help here).

Since Travis CI is dead we must move our CI pipeline to GitHub Actions. I hope I'll find some time in the next couple of days (again... :unamused:) and that it doesn't take too much time. And, of course, Pico 3.0 needs testing...

If you wanna help you can draft the upgrade news (like https://picocms.org/in-depth/upgrade-pico-21/), primarily targeting the changes in our dependencies:

* Twig: We're updating from v1.0 to v3.0, i.e. https://twig.symfony.com/doc/1.x/deprecated.html and https://twig.symfony.com/doc/2.x/deprecated.html both apply
* Symfony YAML: We're updating from v2.8 to v5.0, see https://github.com/symfony/yaml/blob/5.4/CHANGELOG.md
* Parsedown: We're switching back to v1.7 since v1.8 never got final, even though the changes should be negligible (as in https://picocms.org/in-depth/upgrade-pico-21/, just the other way round)

None of the changes should be a big issue, but nevertheless, we should document them... Naturally we also need the common stuff like how to upgrade, but I guess you could simply copy this. Thanks :+1:

mayamcdougall commented:
> Nah, no worries, it's fine. This needs to be done and a little kick in the ass helps 😅 🙈

Alright, lol. And ditto, it usually helps me too. 😉

(But I also know you're usually busy.)

If I can help it, I don't like to leave questions on the Issue tracker unanswered... even if I don't actually *have* an answer.  Sometimes you answer comments right away... and sometimes it takes you a week, lol, and I don't like leaving people in limbo in the meantime.

I've been meaning to add that PHP 8 warning to the readme for awhile, but I've been lazy about it too. 😂

> Since Travis CI is dead we must move our CI pipeline to GitHub Actions.

I've literally never used, or even looked into either option, so I've got no idea about it. (Brief googling looks like they went proprietary recently, so that's not great. 👎🏻)

> If you wanna help you can draft the upgrade news

Yeah, I can take care of that.  Should be easy. 😉

I'll get started on that in a little bit.  Do you have a time frame for when you think you'll be ready with 3.0?  Just a rough idea, eg, "tonight", "this week", or "later on".  For my own fire-lighting, ass-kicking purposes I mean. 😂

mayamcdougall commented:
Writing a note about how YAML now requires strings that start with `%` to be quoted and how that effects Pico users (eg urls starting with `%base_url%`, etc).

Then I realized my example was one of the things needing to be fixed.

https://github.com/picocms/Pico/blob/master/content-sample/_meta.md

So mentioning it here so we don't forget to check that.

mayamcdougall commented:
For the parsedown reversion, do you have any specific examples of what to be on the lookout for?  In the old update post, you only mention issues caused by "non-standard markup", but never go into detail.

(I see you were contributing over there at one point.  I stumbled across it while looking for changes, lol.)

Or should it just be a general, "the version has changed so make sure to recheck that your content files are still rendering correctly" sort of thing?

mayamcdougall commented:
https://github.com/picocms/picocms.github.io/pull/54

This is what I've got so far.  No need to merge yet unless you think I got it perfect in one go, lol. 😂 

Just let me know what you want changed.

It's up on [GitHub Pages](https://mayamcdougall.github.io/picocms.github.io/in-depth/upgrade-pico-30/) *for now*, but I'll be switching GH Pages back to my Docs Rewrite branch once we're done. 😉

mayamcdougall commented:
I've been testing some of the themes against the "Pico 3.0-alpha" build.

I'm up to the letter "D" so far.

I'm starting to see the rationale behind tagging these with some sort of "verified" date. 🤦🏻‍♀️

Still not gonna do it.  But I *am* starting to think that a few of these themes shouldn't make the cut.

I'm discovering a lot more bugs this time around.  Not 3.0 bugs either, but 2.0 ones or just things that were otherwise poor coding or bad practices.

So far the only 3.0 bug was on Astral, which is mine.  Twig changed `divisibleby(x)` to `divisible by(x)`.  No big deal... though I'm seriously bothered by the fact that it has a space now.  It's still a function, not a keyword... but it has a *space*. 😣

WHY?

\**ahem\**

But anyway... I fully expect that by the time I've tested everything, it'll *only* be my work with 3.0 bugs.

Because I would be the one to use obscure Twig functions, wouldn't I? 😂

I'm 100% expecting Notepaper to break, but I haven't gotten there yet.  Maybe that'll finally motivate me to rewrite it. 😒

I'm also taking notes while I go though, for future Theme tagging (R2G, etc). 😉

PhrozenByte committed:
Update various links

PhrozenByte committed:
Add GitHub workflow 'test'

Runs `composer install` and `phpcs` on basically every push, replaces Travis CI.

PhrozenByte committed:
PHP_CodeSniffer: Allow using strftime()

PhrozenByte committed:
GitHub 'test' workflow: Run on any pull request

PhrozenByte committed:
CI: Move install steps to reusable 'install' action

PhrozenByte committed:
CI: Improve reusable 'install' action

PhrozenByte committed:
CI: Update Stale Bot config

PhrozenByte committed:
CI: Improve GitHub Actions YAML formatting

PhrozenByte committed:
CI: Improve Composer root version guessing for PRs

PhrozenByte committed:
Update Pico::VERSION to '3.0.0-dev'

PhrozenByte committed:
Add 'SECURITY.md' to .htaccess

PhrozenByte committed:
phpDoc: Ignore .github dir

PhrozenByte committed:
CI: Check Pico version on tag push

PhrozenByte committed:
CI: Add 'deploy-branch.yml' workflow to update website on branch updates

PhrozenByte committed:
Merge branch 'pico-3.0-alpha' into pico-3.0

PhrozenByte committed:
:tada: Switch from Travis CI to GitHub Actions for release deployment

Since Travis CI was transformed into a proprietary service we didn't have a CI pipeline. This heavily refactors and simplifies the CI pipeline and uses GitHub Actions to deploy new Pico releases. You can use Pico's Makefile to build new releases locally, too. Pico's build script no longer depends on any external resources (like our `picocms/ci-tools` collection). However, this isn't true for other release deployment steps, like updating phpDoc class docs, version badge, cloc stats, etc., even though the CI scripts are mostly self-containing now.

PhrozenByte committed:
CI: Add Composer oauth token for github.com

PhrozenByte committed:
CI: Run 'test' workflow as part of the 'deploy-branch' workflow

Mimics the behaviour of the 'deploy-release' workflow, even though branch deployment doesn't depend on Pico's tests to succeed. Pull requests still use the 'test' workflow directly.

PhrozenByte committed:
composer.json: Allow using all Composer plugins in the 'picocms/*' namespace

PhrozenByte commented:
Here we go... :rocket: :+1:

Sorry for the late response, but I didn't want to answer until this was finished - and it took me longer than expected... :unamused: Anyway, the CI pipeline is back online, even though currently not working here because there are some GitHub repo config settings pending, but it does work both locally and on my fork.

You can give it a try: Simply run `./build/build.sh` locally.

Making this work locally was a major task, but GitHub Actions allows us to remove a whole lot of workaround and helper scripts we used to require on Travis CI. Also see https://github.com/PhrozenByte/Pico/actions/runs/1907606936 and https://github.com/PhrozenByte/Pico/releases/tag/v3.0.0-alpha.3 (no official release, it's just for testing).

There are just three more things to do: (1) Code-wise there are some smaller code changes pending (including the bugs you opened some time ago), but no big issues here. (2) Pico 3.0 needs testing, most notably `PicoDeprecated` and Pico CMS for Nextcloud. (3) Last but not least the upgrade docs.

Thanks for your work on the upgrade docs! :heart: I didn't review your draft yet, just took a quick look. It's too late for today, but I'll likely find some time next weekend :+1:

> I'll get started on that in a little bit. Do you have a time frame for when you think you'll be ready with 3.0? Just a rough idea, eg, "tonight", "this week", or "later on". For my own fire-lighting, ass-kicking purposes I mean. :joy:

It will surely take a few more weeks. I don't want to publish v3.0.0 right away, we should release a beta first and give people 2+ weeks for testing first. Besides, there's still some work to do (see above).

> If I can help it, I don't like to leave questions on the Issue tracker unanswered... even if I don't actually _have_ an answer. Sometimes you answer comments right away... and sometimes it takes you a week, lol, and I don't like leaving people in limbo in the meantime.

Just to explain this: I made the decision that I need to better "economize" with the little time I have for Pico. I can either work on Pico's code, or give extensive support, hardly both. Even though I try to answer any valid question and give at least some quick advice, I decided that I no longer want to do the questioner's job. People need to do their own research and try stuff for themselves. There are basically two possibilities why I don't answer right away: (1) I looked at it and I can't answer it within a few seconds in-between other work - so it's a matter of time; (2) I believe that I already gave enough advice. The latter includes people asking for stuff I believe is no matter of Pico, like webserver config. I still give quick hints, but these projects have their own support channels. The only reason I followed up on the latest issues was because I'll surely always answer *your* questions :+1: (even though not necessarily right away, there's still the time issue :see_no_evil:).

Anyway, please don't read this as some sort of criticism on how you handle things. I truly :heart: and admire the heart and soul you put into answering all questions :+1: I just wanted to explain myself here.

> For the parsedown reversion, do you have any specific examples of what to be on the lookout for? In the old update post, you only mention issues caused by "non-standard markup", but never go into detail.

It's the same as before, just the other way round (i.e. we're going back to what we had before). So I don't expect any major issues from Parsedown, we should just mention it.

> Still not gonna do it. But I am starting to think that a few of these themes shouldn't make the cut.

That's your métier, so I'm fine with whatever you decide :+1:

> I realized the alpha build was only using Twig 2 and not 3

Yes :see_no_evil: From my experience while developing this, I'd say that Symfony YAML now requiring to quote values starting with a `%` (like `Logo: %theme_url%/img/logo.svg`), and Twig removing the `{% for … if … %}` syntax, are the two major issues of Pico 3.0. I know, this is bad :unamused: But we don't really have a choice, Twig 2 is end-of-life, too. If we don't update to Twig 3 we will likely run into similar issues soon.

mayamcdougall commented:
> Sorry for the late response, but I didn't want to answer until this was finished

It's okay, I've had my own testing instance that I cobbled together off the various 3.0 branches. 😂

I overwhelmed myself a little working on the Themes at first, so I've been taking it slower this week.  It's not difficult, just time consuming. 😅

> I made the decision that I need to better "economize" with the little time I have for Pico. / Anyway, please don't read this as some sort of criticism on how you handle things.

Don't worry, I get it. ❤️

Heck, I get overwhelmed with it sometimes and I don't even *have* that many other obligations.  It's mostly just *awkward* at times.  When I have to be like "Yep, idk when he'll be by to answer.  Could be in 5 minutes, could be 5 days. 😅". No worries though, lol.  Besides, it sounds like we're still among the *faster* turnaround times on GitHub.  The impression I've gotten from a few issues has been "It's okay, I expected to be waiting for weeks." 😂

>  I truly ❤️ and admire the heart and soul you put into answering all questions.

Thanks. 😜

As I said recently on that crazy long conversation thread you unsubscribed from\*:

> I like helping people. One reason I provide Pico tech support here is because it's one of the few things I do have. I'm *not* the most knowledgeable. I find it kind of laughable that I'm often the main person answering questions on here and I don't know a scrap of PHP. Talk about imposter syndrome. 😅

> But what keeps me doing it is that, despite not being the best, and despite the imposter syndrome... I'm still the one who can help them. It's something I can do. It's a difference that I can make. And often, I'm the only one who can. If someone's going to provide help... it's me. And that feeling of being *the person* who can make a difference keeps me motivated.

(\* Just describing it that way to avoid linking/tagging.  It's not remotely relevant to the actual discussion here, so I at least *try* not to pollute the repo with unnecessary links, etc.  Also, didn't realize it was a *two-paragraph* quote when I decided to mention it... 😅)

> So I don't expect any major issues from Parsedown, we should just mention it.

It's mentioned.  That's about all I could say on it though, since as far as I can tell, the only changes are obscure bug fixes, edge cases, and such.

> That's your métier, so I'm fine with whatever you decide 👍

Should probably have a proper discussion on it at some point, but the overview is this:  We've got a few themes that are really more of *frameworks* than Themes (Boney, Bootstrap, Bulma, etc).  But, like everything else, they've never been updated.  Bootstrap, for example, is based on Bootstrap 3.0.

It's probably more a naming issue I have with them, as some of them do try to actually *theme*... they're just un-creatively named after framework they used, and I feel that gives the wrong impression.  And on the opposite side, we have specifically themes like "Boney" and "Default Blog", that attempt to recreate the Default Theme, but with different frameworks or capabilities.  I feel like that dilutes things a little.  I wouldn't want users mistaking these for Official, *or* mistaking them for being based on the *current* codebase of the Default Theme.  Maybe I'll just see about pushing some new name / color choices when I eventually PR them. 🤔

Anyway, that's already more than I meant to summarize here. 😒

> I'd say that Symfony YAML now requiring to quote values starting with a `%` (like `Logo: %theme_url%/img/logo.svg`), and Twig removing the `{% for … if … %}` syntax, are the two major issues of Pico 3.0.

I wrote the Upgrade doc before testing Twig 3, but I definitely plan to point out that Twig change (and the YAML issue is already in there).

I did have the realization though: So what you really mean is, every Theme that's ever followed our instruction to borrow from the Default Theme will have issues? 🤦🏻‍♀️

> If we don't update to Twig 3 we will likely run into similar issues soon.

Better to get the breaking done all at once.  I'm not worried about the Theme breakage, and fortunately it sounds like there won't actually be any major Plugin incompatibility, so that's good.

It's also good that Pico doesn't offer any kind of auto-updating, so nothing's going to break without warning.  As far as breaking updates goes, it should be pretty tame.  The biggest issue is going to be everyone who's created or customized their own Theme.  *But* that should hopefully mean that they have the know-how to sort out their own issues, especially if there's a warning upfront in the upgrade doc about what to expect.

I know that's having *way* too much faith in people, but oh well.  We'll never actually hear about the amount of successful upgrades, only the failed ones. 😂

PhrozenByte committed:
Enable PHP strict typing

We can't enable strict typing everywhere without major BC breaks (likely Pico 4.0), so we're doing this on a best-effort basis.

Fixes #603

PhrozenByte committed:
Simplify PHP class imports

PhrozenByte committed:
Proberly handle page data of hidden pages when accessing such a page

This is a workaround for meta pages (i.e. pages starting with a '_'): If a user attempts to request such a page, Pico won't respond with the contents of this meta page, but with a 404 page. This is expected behavior. However, we also have a shortcut in Pico::readPages() attempting to skip reading the contents of the requested page twice. Since we're not serving the contents of the meta page, but of the 404 page, we accidentally overwrite the contents of the meta page by Pico's 404 page. This is unexpected behavior. Even though this commit fixes this particular issue, it doesn't fix its major cause, as the shortcut still exists and can still be triggered by plugin authors by simply overwriting the contents of an existing page. Even though a plugin author might want this to happen, we can't really tell whether it is intended or not. The solution is to remove the shortcut, but we don't want that either, it's a useful performance optimization. The only real solution to this is to switch to page objects, allowing us to handle such situations more verbose. This feature is expected for Pico 4.0. For now we leave this partially fixed...

Fixes #602

PhrozenByte committed:
DummyPlugin: Remove missed imports of the picocms namespace

PhrozenByte committed:
DummyPlugin: Encourage devs to use PHP strict typing

PhrozenByte committed:
CI: Fix Composer root version guessing

PhrozenByte committed:
CI: Some small styling improvements

PhrozenByte committed:
DummyPlugin: Add PHP return types to event methods

Follow-up to 070714d1d82e894b3386f8716e147461e3013098

PhrozenByte committed:
CI: Fix Composer root version guessing (again)

PhrozenByte committed:
CI: Use build script to publish new releases

Since Pico depends on PicoDeprecated, and PicoDeprecated depends on Pico, a chicken-egg-problem arises: When pushing a new tag for Pico, GitHub Actions will try to build this new release by downloading the matching version of PicoDeprecated from Packagist. Since we didn't release a new version of PicoDeprecated yet, this will fail. So we need to push PicoDeprecated's tag first. Doing so yields GitHub Actions trying to build PicoDeprecated's new release - by trying to download the matching version of Pico from Packagist, which doesn't exist yet.

Thus builds will always fail - unless you push both tags at the exact same time, and unless you send some prayers that GitHub Actions delays building until Packagist picks up the new releases. The simplest solution is not to use GitHub Actions to publish new releases, but letting the user to do it instead. With this changeset this is as simple as `make publish version=v3.0.0`. The 'deploy-release' GitHub workflow now just updates the website and is triggered by publishing the GitHub release, not the tag. This additionally allows the user to perform some manual tests before publication.

The build script now depends on GitHub CLI (`gh`), see https://cli.github.com/. You must setup authentication first (`gh auth login`).

PhrozenByte committed:
Add some "art" to Pico class docs

PhrozenByte committed:
Minor code improvements

mayamcdougall commented:
Ooo, some "art". lol. 😉👍🏻

Feels like a good time to mention that I also noticed the credit you added to yourself at the bottom of the Default Theme. 😉

Very well deserved, lol.  I don't have any issue with it, I just noticed it a while back and wanted to be like "Aha, I see what you did there!". 😂

PhrozenByte committed:
Update CONTRIBUTING.md to include latest changes

PhrozenByte commented:
"CI is ready" was a bit premature, we kinda had a little chicken-and-egg situation with automatically building Pico and PicoDeprecated: Pico depends on PicoDeprecated to build, and PicoDeprecated depends on Pico to build. Fixed with [60d0f54](https://github.com/picocms/Pico/pull/535/commits/60d0f5403c7b7fd00b068e8ede3326a3b566bcdd). Builds might still fail if tags aren't pushed at the same time, but that's no big issue because one can simply manually start the workflow another time, check the result and then manually push a new release with `make publish`. Also see [81c2172](https://github.com/picocms/Pico/pull/535/commits/81c21722d32a37458b26a4434531a2ef016e272f)

> I overwhelmed myself a little working on the Themes at first, so I've been taking it slower this week. It's not difficult, just time consuming. sweat_smile

No worries, it isn't ready for release yet anyway :wink:

> Heck, I get overwhelmed with it sometimes and I don't even _have_ that many other obligations. It's mostly just _awkward_ at times.

:heart: :+1:

> Should probably have a proper discussion on it at some point, but the overview is this: We've got a few themes that are really more of _frameworks_ than Themes (Boney, Bootstrap, Bulma, etc). But, like everything else, they've never been updated. Bootstrap, for example, is based on Bootstrap 3.0.

If it's this old and you can't reach anyone to inform the dev that the theme needs an update, we should remove it. I'm fine with removing more-or-less broken, or heavily outdated themes (and plugins, even though we didn't merge the wiki yet) which are no longer supported by their respective devs. You shouldn't feel obligated to do their work for them, if they don't want to update their themes, simply remove them.

> I did have the realization though: So what you really mean is, every Theme that's ever followed our instruction to borrow from the Default Theme will have issues? 🤦🏻‍♀️

Unfortunately yes :unamused: 

> It's also good that Pico doesn't offer any kind of auto-updating, so nothing's going to break without warning. As far as breaking updates goes, it should be pretty tame. The biggest issue is going to be everyone who's created or customized their own Theme.

So, everyone :laughing: But yes, it shouldn't be a big issue for most users. I assume that a lot users will stick to Pico 2.1 at first, at least until their respective web hoster drops PHP 7 support. The biggest issue is that we likely loose a lot (abandoned) themes. However, personally I don't value abandonware much :wink:

> Ooo, some "art". lol. wink👍🏻
> 
> Feels like a good time to mention that I also noticed the credit you added to yourself at the bottom of the Default Theme. wink

:wink:

> [picocms/picocms.github.io#54](https://github.com/picocms/picocms.github.io/pull/54)
> 
> This is what I've got so far. No need to merge yet unless you think I got it perfect in one go, lol. joy
> 
> Just let me know what you want changed.
> 
> It's up on [GitHub Pages](https://mayamcdougall.github.io/picocms.github.io/in-depth/upgrade-pico-30/) _for now_, but I'll be switching GH Pages back to my Docs Rewrite branch once we're done. wink

I'll review the changes there.

PhrozenByte committed:
Update CHANGELOG.md

PhrozenByte committed:
Update dependencies in composer.json

PhrozenByte committed:
Remove Pico's `map` Twig filter

Fixes #623

PhrozenByte committed:
Sample contents: Remove Pico's `map` Twig filter

PhrozenByte committed:
Update sample contents for Pico 3.0

PhrozenByte committed:
Update README.md and sample contents for Pico 3.0

PhrozenByte committed:
Add 'page' Twig function to return a single page

PhrozenByte committed:
Allow $markdown = null in 'markdown' Twig filter

PhrozenByte committed:
Various small improvements

PhrozenByte committed:
CI: Add bouncer step to PR testing

dkyme committed:
replace strlen in readPages() with variables

see issue #671

PhrozenByte committed:
Merge pull request #674 from dkyme/dkyme-patch-1

replace strlen in readPages() with variables

0xMax42 committed:
Enhanced Sorting by Nested Meta Values in Pages (#681)

* Added functionality for sorting pages by nested meta values.

* Additional variables removed.

* Added comment to the config template file about the possibility to access nested metadata with a dot as separator for sorting.

* Added entry in the changelog about the possibility to access nested metadata with a dot as separator for sorting.

* Changelog and config template adapted according to the suggestions.

* Added old entry back in the Config template.

dkyme committed:
fix Wrong config keys in discoverPageSiblings #675

PhrozenByte committed:
Merge pull request #686 from dkyme/fix-discoverPageSiblings

fix Wrong config keys in discoverPageSiblings #675

PhrozenByte committed:
Pico::readPages(): Improve variable naming

PhrozenByte committed:
Merge branch 'master' into pico-3.0

PhrozenByte committed:
Update CI workflows to test from PHP 7.2 to 8.5

PhrozenByte committed:
Fix CI deployment workflows

PhrozenByte committed:
Update CI deployment workflows to run with PHP 8.5 and latest phpDocumentor

PhrozenByte committed:
Add CI workflow to build release packages

PhrozenByte committed:
Allow Composer plugin `picocms/composer-installer` in composer.json

PhrozenByte committed:
Add `/Makefile export-ignore` to .gitattributes

PhrozenByte committed:
Support building non-dev versions in build workflow

PhrozenByte committed:
Enable build workflow for pico-3.0 branch

PhrozenByte committed:
Fix nullable type declarations

PhrozenByte committed:
Update Twig and Symfony dependencies

Choosing the latest version that still supports PHP 7.2, but includes the most fixes of later PHP deprecations.

PhrozenByte committed:
Disable E_DEPRECATED, E_USER_DEPRECATED, and E_STRICT error reporting

Some of our dependencies weren't made for late PHP 8.x versions, thus yielding many deprecation and strict errors. Let's hide them by default. We really have to call `error_reporting()` twice, because `E_STRICT` was deprecated too, otherwise yielding an deprecation notice.

PhrozenByte committed:
Enable GitHub release deployment for build workflow
